[ADD] extensions: new autofield directive, [REF] i18n/localization.rst (#1334)
odoo/documentation: content/developer/howtos/localization.rst

            "data/account.account.template.csv",
            "data/account.group.template.csv",

            # Taxes
            "data/account_tax_group_data.xml",
            "data/account_tax_report_data.xml",
            "data/account_tax_template_data.xml",
            "data/account_fiscal_position_template_data.xml",
            "data/account_account_template_post_data.xml",

            "data/account_chart_post_data.xml",
            "data/account_chart_template_try_loading.xml",

            # Views and others
            "views/xxxmodel_views.xml"
        ],
        "demo": [
            "demo/demo_company.xml",
        ]
    }

In the first file `data/account_chart_template_data.xml` we give name for the Chart of Accounts and set some basic fields.

.. seealso::
   :ref:`Chart Template References <reference/account_chart_template>`

.. example::
  `addons/l10n_ch/data/l10n_ch_chart_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/l10n_ch_chart_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/l10n_ch_chart_data.xml
    :language: xml
    :start-at: l10nch_chart_template
    :end-at: </record>

.. note::

>   Recommended **xmlid** for the record is `chart_template`.

Reviewer: Funny that the example above doesn't respect that :p Should we even bother recommanding anything here ? It actually makes no difference from a technical point of view.

Author: It is annoying to have `l10n_xx.l10nxx_chart_template`. (quite repetitive) Would be easier if it was uniform.

  If you need many chart of accounts, you can add some suffixes, i.e. `chart_template_XXX`.

Chart of Accounts
=================

Accounts
--------

.. seealso::
   - :ref:`Account References <reference/account_account>`
   - :doc:`/applications/finance/accounting/payables`

>    - :doc:`/applications/finance/accounting/receivables`

Reviewer: Why linking the doc about payable and receivable accounts here? To me, they have nothing to do with the creation of an l10n module/CoA.

Author: They help devs to understand the account types when they have to chose the right ones

Reviewer: Hmmm, not convinced; to me it's giving a little too much details regarding what we are explaining here.

Author: I changed the link to the doc, maybe you'll find it more relevant

Obviously, *Chart of Accounts* cannot exist without *Accounts*. You need to specify them in `data/account.account.template.csv`.

.. example::
  `addons/l10n_ch/data/account.account.template.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/account.account.template.csv>`__.

>   .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/account.account.template.csv

Reviewer 2: would probably present better with a csv-table + file directive. Can you use start at/end-at as well in that?

Author: Not that easy https://docutils.sourceforge.io/docs/ref/rst/directives.html#csv-table

I tried with a `.. include` and `.. csv-table` but without success

    :language: csv
    :end-at: ch_coa_1171

> XML is prefered but you may use xml format instead.

Reviewer 3: Did you mean CSV is better ?

> XML is prefered but you may use xml format instead.

Reviewer: csv 
```suggestion
csv is preferred but you may use xml format instead.
```

    :language: xml
    :start-at: chart_at_template_0010
    :end-at: </record>

.. warning::

    Avoid the usage of liquidity ``account.account.type``!
    Indeed, the bank & cash accounts are created directly at the installation of the localization module and then, are linked to an ``account.journal``.

.. warning::

    Only one account of type payable/receivable is enough.

.. warning::

    Don't create too much accounts: 200-300 is enough.

Next settings for the chart of accounts are set in a separate file, because we need to provide `list of accounts <#accounts>`__ first. In `data/account_chart_post_data.xml` we set some default accounts:

.. todo add reference to account_id in CoA

.. example::
  `addons/l10n_ch/data/l10n_ch_chart_post_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/l10n_ch_chart_post_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/l10n_ch_chart_post_data.xml
    :language: xml
    :start-at: l10nch_chart_template
    :end-at: </record>

Account tags
------------

.. seealso::
   :ref:`Account Tag References <reference/account_account_tag>`

> Tags is a way to sort accounts.

Reviewer: ```suggestion
Tags are a way to sort accounts.
```

For example, imagine you want to create a financial report having multiple lines but you have no way to find a rule to dispatch the accounts according their ``code`` or ``name``.
The solution is the usage of tags, one for each report line, to filter accounts like you want.

Put the tags in `data/account_account_tag_data.xml` file:

> Then apply tags on some accounts (field `tag_ids`).

Reviewer: ```suggestion
Put the tags in `data/account_account_tag_data.xml` file, then apply the tags to some accounts (field `tag_ids`).
```

.. example::
  `addons/l10n_lt/data/account.account.template.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_lt/data/account.account.template.csv>`__.

>   .. literalinclude:: {ODOO_RELPATH}/addons/l10n_lt/data/account.account.template.csv

Reviewer 2: would probably present better with a `csv-table + file` directive. Can you use `start at`/`end-at` as well in that?

    :language: csv
    :end-at: account_account_template_1201

.. example::
  `addons/l10n_at/data/account_account_template.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_at/data/account_account_template.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_at/data/account_account_template.xml
    :language: xml
    :start-at: chart_at_template_0010
    :end-at: </record>

Account groups
--------------

.. seealso::
   :ref:`Account Group References <reference/account_group>`

> Account groups allows to describe structure of chart of accounts.

Reviewer: ```suggestion
Account groups allow describing the structure of a chart of accounts.
```

.. example::
  `addons/l10n_il/data/account.group.template.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_il/data/account.group.template.csv>`__.

  .. csv-table::
     :file: {ODOO_ABSPATH}/addons/l10n_il/data/account.group.template.csv
     :widths: 20,20,20,20,20
     :header-rows: 1

Taxes
-----

.. seealso::
   - :ref:`Tax References <reference/account_tax>`
   - :doc:`/applications/finance/accounting/taxation/taxes/taxes`

> To add taxes you first need to specify tax groups. It has just two fields: name and country. Create file `data/account_tax_group_data.xml` and list the groups:

Reviewer: ```suggestion
To add taxes you first need to specify tax groups. This model only has two required fields: name and country. Create file `data/account_tax_group_data.xml` and list the groups:
```

.. code-block:: xml

    <?xml version="1.0" encoding="utf-8"?>
    <odoo>
        <data noupdate="1">
            <record id="tax_group_tva_0" model="account.tax.group">
                <field name="name">TVA 0%</field>
                <field name="country_id" ref="base.ch"/>
            </record>

            ...
        </data>

>     </odoo>

Reviewer: I don't think you need this with the example below.

.. example::
  `addons/l10n_ch/data/account_tax_group_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/account_tax_group_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/account_tax_group_data.xml
    :language: xml
    :start-after: <data
    :end-before: </data>

.. example::
  `addons/l10n_uk/data/account.tax.group.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_uk/data/account.tax.group.csv>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_uk/data/account.tax.group.csv

>     :language: csv

Reviewer 2: idem

Now you can add the taxes via `data/account_tax_template_data.xml` file.

.. example::
  `addons/l10n_ae/data/account_tax_template_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ae/data/account_tax_template_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ae/data/account_tax_template_data.xml
    :language: xml
    :start-at: uae_sale_tax_5_dubai
    :end-at: </record>

> If some accounts should use default taxes, you can set them up in `data/account_account_template_post_data.xml`

Reviewer: Maybe link an example ?

Author: The only case where it's done doesn't follow this guideline, in l10n_de (and makes the code a mess)

Tax Report
----------

.. seealso::
   - :ref:`Tax Report Line References <reference/account_tax_report_line>`

>    - :doc:`/applications/finance/accounting/reporting/declarations/tax_returns`

Reviewer: On financial reports, we explicitly state they are enterprise; we should do the same here. Only the model and the definition of its instances are community. Maybe just explaining it with 1-2 sentences here would be nice as well.

> In previous section you noticed fields `invoice_repartition_line_ids` / `refund_repartition_line_ids` and probably understand nothing about it. Good news: you are not alone on this incomprehension. Bad news: you have to figure it out a bit. The topic is complicated indeed:

Reviewer: ```suggestion
In the previous section you noticed fields `invoice_repartition_line_ids` / `refund_repartition_line_ids` and probably understood nothing about them. Good news: you are not alone on this incomprehension. Bad news: you have to figure it out a bit. The topic is complicated indeed:
```

> .. graphviz:: images/tax_report.dot

Reviewer: It's quite small on my screen and there doesn't seem to be a way of maximizing it ... can we do something ?

Author: Is it better now?

> But lucky we have a presentation explaining the tax reports in details:

Reviewer: ```suggestion
But fortunately we have a presentation explaining the tax reports in details:
```

.. youtube:: PuXE_NyFRTM
    :align: right
    :width: 700

>     :height: 394

Reviewer: As I said on the other PR, I'm not sure it's a good idea to link it here. This was done for v13, and the model has changed a little since then. Moreover, it essentially compares the financial reports with the new engine, and I'm not sure it'll help newcomers in any way.

Author: It is indeed explaining the difference (we don't care about that part), but also explaining how the new model works (we care about this)

Reviewer: But on the other hand, it's already partly outdated, as account.tax.report objects didn't exist back then. Running exactly the examples from the presentation won't work. I'm not 100% against keeping it, but we need to keep that in mind. And I'm not sure we can say it explicitly here either, like "oh, we know we're giving you something a bit outdated, but who cares, we didn't want to make a new one" :p Maybe there is a more diplomatic way, though.

> So, once you have properly configured taxes, you just need to add `data/account_tax_report_data.xml` file with a single record for `account.tax.report` in beginning:

Reviewer: ```suggestion
So, once you have properly configured taxes, you just need to add `data/account_tax_report_data.xml` file with a record for your `account.tax.report` at the beginning:
```

.. code-block:: xml

    <?xml version="1.0" encoding="utf-8"?>
    <odoo>
        <record id="tax_report" model="account.tax.report">
            <field name="name">Tax Report</field>
            <field name="country_id" ref="base.XX"/>
        </record>

        ...
    </odoo>

> ... following by a list of records for `account.tax.report.line`.

Reviewer: ```suggestion
... followed by the declaration of its lines, as `account.tax.report.line` records.
```

.. example::
  `addons/l10n_au/data/account_tax_report_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_au/data/account_tax_report_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_au/data/account_tax_report_data.xml
    :language: xml
    :start-at: tax_report
    :end-before: account_tax_report_gstrpt_g3

Fiscal positions
----------------

.. seealso::
   - :ref:`Fiscal Position References <reference/account_fiscal_position>`
   - :doc:`/applications/finance/accounting/taxation/taxes/fiscal_positions`

> Specify fiscal positions in `data/account_fiscal_position_template_data.xml` file as following:

Reviewer: ```suggestion
Specify fiscal positions in `data/account_fiscal_position_template_data.xml` file.
```

> You may have as many fiscal position as you need.

Reviewer: Kind of obvious. I wouldn't say it.

.. example::
  `addons/l10n_es/data/account_fiscal_position_template_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_es/data/account_fiscal_position_template_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_es/data/account_fiscal_position_template_data.xml
    :language: xml
    :start-at: fp_nacional
    :end-before: fp_intra

Final Steps
===========

> The last step on installing Chart of Accounts is applying it on current company. File `data/account_chart_template_try_loading.xml` is responsible for that.

Reviewer: ```suggestion
The last step when installing a localization module is to try applying its chart of accounts to the current company (if it does not already have one). 
File `data/account_chart_template_try_loading.xml` is responsible for that.
```

=======================
Accounting Localization
=======================

.. warning::

    This tutorial requires knowledges about how to build a module in Odoo (see
    :doc:`/developer/howtos/backend`).

How localization works
======================

On installing the `account <https://github.com/odoo/odoo/tree/15.0/addons/account>`__ module, the localization module corresponding to the country code of the company is installed automatically.
In case of no country code set or no localization module found, the `l10n_generic_coa <https://github.com/odoo/odoo/tree/15.0/addons/l10n_generic_coa>`__ (US) localization module is installed by default.

> Check `post init hook <https://github.com/odoo/odoo/blob/15.0/addons/account/__init__.py>`__ for details.

Author: It doesn't target devs. And trust me, it is not that easy to find.

To quote you:
> Did you ask newbies in your team to give their POV on this doc?
As developers, we easily believe everything is easy to understand in our own scope and to oversimplify the our explanations.

And yes. I asked newbies.

> Check `post init hook <https://github.com/odoo/odoo/blob/15.0/addons/account/__init__.py>`__ for details.

Reviewer 4: If it doesn't target devs, why is it in the `developer/` section then ?

> Check `post init hook <https://github.com/odoo/odoo/blob/15.0/addons/account/__init__.py>`__ for details.

Author: It targets devs, but some are more "hardcore" than others. We often have PRs of ppl who can barely use a version control system for instance. Or devs du dimanche who want to help building a localization for their country, who know how to write xml but are afraid of digging in the code.

            "data/account_account_tag_data.xml",
            "data/account.account.template.csv",
            "data/account.group.template.csv",

            # Taxes
            "data/account_tax_group_data.xml",
            "data/account_tax_report_data.xml",
            "data/account_tax_template_data.xml",
            "data/account_fiscal_position_template_data.xml",
            "data/account_account_template_post_data.xml",

            "data/account_chart_post_data.xml",
            "data/account_chart_template_try_loading.xml",

            # Views and others
            "views/xxxmodel_views.xml"
        ],
        "demo": [
            "demo/demo_company.xml",
        ]
    }

In the first file `data/account_chart_template_data.xml` we give name for the Chart of Accounts and set some basic fields.

.. seealso::
   :ref:`Chart Template References <reference/account_chart_template>`

.. example::
  `addons/l10n_ch/data/l10n_ch_chart_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/l10n_ch_chart_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/l10n_ch_chart_data.xml
    :language: xml
    :start-at: l10nch_chart_template

>     :end-at: </record>

Author: Will be outdated sooner than later then. We don't care if there are no examples for ppl who build the doc without the source code.
I want the doc to raise a warning if the source code is available but the `start-at` raises a warning for instance.

>     :end-at: </record>

Author: Fixed using a `condition` option

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_lt/data/account.account.template.csv
    :language: csv
    :end-at: account_account_template_1201  

.. example::
  `addons/l10n_at/data/account_account_template.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_at/data/account_account_template.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_at/data/account_account_template.xml
    :language: xml
    :start-at: chart_at_template_0010
    :end-at: </record>

Accounts
--------

.. seealso::
   - :ref:`Account References <reference/account_account>`
   - :doc:`/applications/finance/accounting/payables`
   - :doc:`/applications/finance/accounting/receivables`

Obviously, *Chart of Accounts* cannot exist without *Accounts*. You need to specify them in `data/account.account.template.csv`.

.. example::
  `addons/l10n_ch/data/account.account.template.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/account.account.template.csv>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/account.account.template.csv
    :language: csv
    :end-at: ch_coa_1171

> CSV is prefered but you may use XML format instead.

Reviewer 4: a little note on the perf diff?

Author: There is none (tested one or two years ago, after 13.0). 
It is just for readability.

Reviewer 4: IIRC, csv records are created in batch, which isn't the case for xml records :thinking:

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_at/data/account_account_template.xml
    :language: xml
    :start-at: chart_at_template_0010
    :end-at: </record>

.. warning::

    Avoid the usage of liquidity ``account.account.type``!
    Indeed, the bank & cash accounts are created directly at the installation of the localization module and then, are linked to an ``account.journal``.

.. warning::

    Only one account of type payable/receivable is enough.

.. warning::

    Don't create too much accounts: 200-300 is enough.

Next settings for the chart of accounts are set in a separate file, because we need to provide `list of accounts <#accounts>`__ first. In `data/account_chart_post_data.xml` we set some default accounts:

.. todo add reference to account_id in CoA

.. example::
  `addons/l10n_ch/data/l10n_ch_chart_post_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/l10n_ch_chart_post_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/l10n_ch_chart_post_data.xml
    :language: xml
    :start-at: l10nch_chart_template

>     :end-at: </record>

Reviewer 5: The example gives 7 fields only, while it's 21 fields

```xml

  <!-- Receivable/payable accounts. -->
  <field name="property_account_receivable_id" ref="..."/>
  <field name="property_account_payable_id" ref="..."/>

  <!-- Default Income/Expense Account for a Product Template. -->
  <field name="property_account_income_id" ref="..."/>
  <field name="property_account_expense_id" ref="..."/>

  <!-- Default Income/Expense Account for a Product Category. -->
  <field name="property_account_income_categ_id" ref="..."/>
  <field name="property_account_expense_categ_id" ref="..."/>

  <!-- Cash loss/gain accounts. They correspond profit_account_id / loss_account_id fields -->
  <field name="default_cash_difference_income_account_id" ref="..."/>
  <field name="default_cash_difference_expense_account_id" ref="..."/>

  <!-- Loss/gain exchange rate accounts. -->
  <field name="income_currency_exchange_account_id" ref="..."/>
  <field name="expense_currency_exchange_account_id" ref="..."/>

  <!-- Stock valuation accounts. -->
  <field name="property_stock_account_input_categ_id" ref="..."/>
  <field name="property_stock_account_output_categ_id" ref="..."/>
  <field name="property_stock_valuation_account_id" ref="..."/>

  <!-- Tax accounts to  balance current/advance tax payments. -->
  <!-- It's used in `env['account.generic.tax.report']._generate_tax_closing_entries` -->
  <field name="property_tax_payable_account_id" ref="..."/>
  <field name="property_tax_receivable_account_id" ref="..."/>
  <field name="property_advance_tax_payment_account_id" ref="..."/>

  <!-- Base Tax Received Account -->
  <!-- Corresponds to `account_cash_basis_base_account_id` field. -->
  <!-- It's used for taxes with `tax_exigibility` equal to `on_payment` -->
  <field name="property_cash_basis_base_account_id" ref="..."/>

  <!-- POS account. -->
  <field name="default_pos_receivable_account_id" ref="..."/>

  <!-- Payment accounts. -->
  <!--
       Suspense account is used to register an unreconiled bank entries created via bank statement.
       Once it's reconciled, the suspense account in `account.move` will be replaced with a proper account.
       Suspense account is an exceptional case when account is supposed to be replaced in the transaction.
       The new account depends on `account.payment` records. There are two cases
       1. `account.payment' exists before reconciliation aka Blue lines in reconciliation wizard
       2. `account.payment' is created on reconciliation aka regular Black lines in reconciliation wizard

       For case n.1 the new account is Outstanding Receipts/Payments account and the workflow is following:

       Initial transactions:
       * Invoice/bill: Payable/Receivable vs Expenses/Sales
       * Payment:             Outstanding vs Payable/Receivable
       After creating bank statement
       * Bank entry:                 Bank vs Suspense
       After reconciliation:
       * Bank entry:                 Bank vs Outstanding

       For case n.2 the new account is Payable/Receivable account from invoice/bill and the workflow is following:

       Initial transactions:
       * Invoice/bill: Payable/Receivable vs Expenses/Sales
       After creating bank statement
       * Bank entry:                 Bank vs Suspense
       After reconciliation:
       * Bank entry:                 Bank vs Payable/Receivable
  -->
  <field name="account_journal_suspense_account_id" ref="..."/>
  <!-- Outstanding Receipts Account -->
  <field name="account_journal_payment_debit_account_id" ref="..."/>
  <!-- Outstanding Payments Account -->
  <field name="account_journal_payment_credit_account_id" ref="..."/>

```

Reviewer 5: # Current version
![image](https://user-images.githubusercontent.com/186131/143901386-1de63d04-cf61-4f60-a2d8-d71d45b4c238.png)

Author: Because you can't put them all at the same time, you first need to create the accounts, look below in the HowTo, where the steps are done in the right order

Reviewer 5: You probably misunderstand me. I'm talking about post-data file, not about the first file `account_chart_template_data`

Author: You have the reference for that, no need to put an example with every single account set. I'm not even sure it exists.

Account groups
--------------

.. seealso::
   :ref:`Account Group References <reference/account_group>`

Account groups allow describing structure of chart of accounts.

.. example::
  `addons/l10n_il/data/account.group.template.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_il/data/account.group.template.csv>`__.

  .. csv-table::
     :file: {ODOO_ABSPATH}/addons/l10n_il/data/account.group.template.csv
     :widths: 20,20,20,20,20
     :header-rows: 1

Taxes
-----

.. seealso::
   - :ref:`Tax References <reference/account_tax>`
   - :doc:`/applications/finance/accounting/taxation/taxes/taxes`

> To add taxes you first need to specify tax groups. This model only has two required fields: name and country. Create file `data/account_tax_group_data.xml` and list the groups:

Reviewer 6: ```suggestion
To add taxes you first need to specify tax groups. You normally need just one tax group for every tax rate, except for the 0% as you need to often distinguish between exempt, 0%, not subject, ... taxes. 
 This model only has two required fields: name and country. Create file `data/account_tax_group_data.xml` and list the groups:
```

.. code-block:: xml

    <?xml version="1.0" encoding="utf-8"?>
    <odoo>
        <data noupdate="1">
            <record id="tax_group_tva_0" model="account.tax.group">
                <field name="name">TVA 0%</field>
                <field name="country_id" ref="base.ch"/>
            </record>

            ...
        </data>
    </odoo>

.. example::
  `addons/l10n_ch/data/account_tax_group_data.xml <https://github.com/odoo/odoo/blob/15.0/addons/l10n_ch/data/account_tax_group_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/account_tax_group_data.xml
    :language: xml
    :start-after: <data
    :end-before: </data>

.. example::
  `addons/l10n_uk/data/account.tax.group.csv <https://github.com/odoo/odoo/blob/15.0/addons/l10n_uk/data/account.tax.group.csv>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_uk/data/account.tax.group.csv
    :language: csv

> Now you can add the taxes via `data/account_tax_template_data.xml` file.

Reviewer 6: ```suggestion
Now you can add the taxes via `data/account_tax_template_data.xml` file.  The first tax you define that is purchase/sale also becomes the default purchase/sale tax for your products.  
```

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ae/data/account_tax_template_data.xml
    :language: xml
    :start-at: uae_sale_tax_5_dubai
    :end-at: </record>

If some accounts should use default taxes, you can set them up in `data/account_account_template_post_data.xml`

Tax Report
----------

.. raw:: html

   <div><span class="badge" style="background-color:#AD5E99">Enterprise feature</span><div>

The tax report is declared in the Invoicing (`account`) app, but the report is only accessible when Accounting (`account_accountant`) is installed.

.. seealso::
   - :ref:`Tax Report Line References <reference/account_tax_report_line>`
   - :doc:`/applications/finance/accounting/reporting/declarations/tax_returns`

In the previous section you noticed fields `invoice_repartition_line_ids` / `refund_repartition_line_ids` and probably understood nothing about them. Good news: you are not alone on this incomprehension. Bad news: you have to figure it out a bit. The topic is complicated indeed:

.. graphviz:: images/tax_report.dot
    :class: overflow-auto

> But fortunately we have a presentation explaining the tax reports (as in version 13.0) in details:

Reviewer 6: @william-andre I have already commented about this in the other PR as well:  

```suggestion
The simple version is that in the tax template, you indicate in the invoice/refund repartition lines whether base or a percentage of the tax needs to be reported in which report line (through the minus/plus_report_line_ids).  It becomes clear also when you check the tax configuration in the Odoo interface (or check the docs (ref))  And fortunately we have a presentation explaining the tax reports (as in version 13.0) in details:
```

=======================

> Accounting Localization

Reviewer 7: My bad, I didn't know about this rule.

Tags are a way to sort accounts.
For example, imagine you want to create a financial report having multiple lines but you have no way to find a rule to dispatch the accounts according their ``code``.
The solution is the usage of tags, one for each report line, to filter accounts like you want.

Put the tags in `data/account_account_tag_data.xml` file.

.. example::
  `addons/l10n_lt/data/account.account.template.csv <{GITHUB_PATH}/addons/l10n_lt/data/account.account.template.csv>`__

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_lt/data/account.account.template.csv
    :condition: odoo_dir_in_path
    :language: csv
    :end-at: account_account_template_1201

.. example::
  `addons/l10n_at/data/account_account_template.xml <{GITHUB_PATH}/addons/l10n_at/data/account_account_template.xml>`__

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_at/data/account_account_template.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-at: chart_at_template_0010
    :end-at: </record>

Accounts
--------

.. seealso::
   - :ref:`Account References <reference/account_account>`
   - :doc:`/applications/finance/accounting/getting_started/initial_configuration/chart_of_accounts`

> Obviously, *Chart of Accounts* cannot exist without *Accounts*. You need to specify them in `data/account.account.template.csv`.

Author: Is there a documentation to where we can find all the roles (like `guilabel` or `file`) that we should use?

> Obviously, *Chart of Accounts* cannot exist without *Accounts*. You need to specify them in `data/account.account.template.csv`.

Reviewer 7: Not *yet*: https://www.odoo.com/web#id=2810468&cids=1&menu_id=5195&action=333&active_id=3835&model=project.task&view_type=form

    :condition: odoo_dir_in_path
    :language: xml
    :start-at: l10nch_chart_template
    :end-at: </record>

Account groups
--------------

.. seealso::
   :ref:`Account Group References <reference/account_group>`

Account groups allow describing the hierarchical structure of the chart of accounts. The filter needs to be activated in the report and then when you decollapse into journal entries it will show the parents of the account.

It works with a prefix start/end, so every account where the code starts with something between start and end will have this account.group as parent group.  Furthermore, the account groups can have a parent account group as well to form the hierarchy.

.. example::
  `addons/l10n_il/data/account.group.template.csv <{GITHUB_PATH}/addons/l10n_il/data/account.group.template.csv>`__

  .. csv-table::
     :condition: odoo_dir_in_path
     :file: {ODOO_ABSPATH}/addons/l10n_il/data/account.group.template.csv
     :widths: 20,20,20,20,20
     :header-rows: 1

Taxes
-----

.. seealso::
   - :ref:`Tax References <reference/account_tax>`
   - :doc:`/applications/finance/accounting/taxation/taxes/taxes`

> To add taxes you first need to specify tax groups. You normally need just one tax group for every tax rate, except for the 0% as you need to often distinguish between exempt, 0%, not subject, ... taxes.

Reviewer 8: Is it useful to mention tax groups' subtotals are displayed below the invoice, but tags are used in tax reports?

Author: Probably, you are welcome to suggest some changes! :)

Reviewer 6: There are plenty ideas for changes, but we thought to have this one merged first

Author: I want to merge it as soon as it is good on the tech pov for everyone.
While it isn't, anyone can already add some functional things, no need to delay that.

.. example::
  `addons/l10n_au/data/account_tax_report_data.xml <{GITHUB_PATH}/addons/l10n_au/data/account_tax_report_data.xml>`__

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_au/data/account_tax_report_data.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-at: tax_report
    :end-before: account_tax_report_gstrpt_g3

Fiscal positions
----------------

.. seealso::
   - :ref:`Fiscal Position References <reference/account_fiscal_position>`
   - :doc:`/applications/finance/accounting/taxation/taxes/fiscal_positions`

Specify fiscal positions in `data/account_fiscal_position_template_data.xml` file.

.. example::
  `addons/l10n_es/data/account_fiscal_position_template_data.xml <{GITHUB_PATH}/addons/l10n_es/data/account_fiscal_position_template_data.xml>`__

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_es/data/account_fiscal_position_template_data.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-at: fp_nacional
    :end-before: fp_intra

Final Steps
===========

The last step when installing a localization module is to try applying its chart of accounts to the current company (if it does not already have one).

> File `data/account_chart_template_try_loading.xml` is responsible for that.

Reviewer 7: ```suggestion
The file :file:`data/account_chart_template_try_loading.xml` is responsible for that.
```

.. example::
  `addons/l10n_ch/data/account_chart_template_data.xml <{GITHUB_PATH}/addons/l10n_ch/data/account_chart_template_data.xml>`__

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/account_chart_template_data.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-at: <function
    :end-at: </function>

Finally, you may add a demo company, so the localization can be easily tested in demo mode.

.. example::
  `addons/l10n_ch/demo/demo_company.xml <{GITHUB_PATH}/addons/l10n_ch/demo/demo_company.xml>`__

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/demo/demo_company.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-after: <odoo>
    :end-before: </odoo>

Accounting reports
==================

.. raw:: html

   <div><span class="badge" style="background-color:#AD5E99">Enterprise feature</span><div>

.. seealso::
  :doc:`/applications/finance/accounting/reporting/overview`

> Accounting reports should be added via a separate module **l10n_XX_reports** that should go to the `enterprise repository <https://github.com/odoo/enterprise>`__.

Reviewer 7: ```suggestion
Accounting reports should be added via a separate module `l10n_XX_reports` that should go to the `enterprise repository <https://github.com/odoo/enterprise>`_.
```

Basic `__manifest__.py` file for such a module looks as following:

.. code-block:: py

    {
        "name": "COUNTRY - Accounting Reports",
        "category": "Accounting/Localizations/Reporting",
        "version": "1.0.0",
        "license": "OEEL-1",
        "depends": [
            "l10n_XX", "account_reports"
        ],
        "data": [
            "data/account_financial_html_report_data.xml",
        ],
        "auto_install": True,
    }

Functional overview of financial reports is here: :doc:`/applications/finance/accounting/reporting/overview/main_reports`.

Some good examples:

> * `l10n_ch_reports/data/account_financial_html_report_data.xml <https://github.com/odoo/enterprise/blob/15.0/l10n_ch_reports/data/account_financial_html_report_data.xml>`__

Reviewer 7: Wouldn't that make sense to have a replacement for enterprise URLs too?

Author: Done

* `l10n_be_reports/data/account_financial_html_report_data.xml <https://github.com/odoo/enterprise/blob/15.0/l10n_be_reports/data/account_financial_html_report_data.xml>`__

> For the fields meaning dive directly to the source:

Reviewer 7: ```suggestion
For the fields' meaning, dive directly to the source:
```

* `account.financial.html.report (v15) <https://github.com/odoo/enterprise/blob/d4eff9d39469cf3fe18589a1547cb0cdb93f4ae9/account_reports/models/account_financial_report.py#L59-L75>`__
* `account.financial.html.report.line (v15) <https://github.com/odoo/enterprise/blob/d4eff9d39469cf3fe18589a1547cb0cdb93f4ae9/account_reports/models/account_financial_report.py#L931-L964>`__

> Menu for the new report is created automatically. By default it's located under *Accounting >> Reporting* menu.

Reviewer 7: ```suggestion
The menu for the new report is created automatically. By default, it is located under :menuselection:`Accounting -> Reporting`.
```

> To create a dedicated section in Reporting menu, you need to create new `ir.ui.menu` record (usually in the main `l10n_XX` module) and set it as `parent_id` field in `account.financial.html.report` model. Example for Belgium localization:

Reviewer 7: ```suggestion
To create a dedicated section in the :guilabel:`Reporting` menu, you need to create a new `ir.ui.menu` record (usually in the main `l10n_XX` module) and set it as `parent_id` field in the `account.financial.html.report` model. Example for the Belgian localization:
```

            # Views and others
            "views/xxxmodel_views.xml"
        ],
        "demo": [
            "demo/demo_company.xml",
        ]
    }

In the first file `data/account_chart_template_data.xml` we give name for the Chart of Accounts and set some basic fields.

.. seealso::
   :ref:`Chart Template References <reference/account_chart_template>`

.. example::
  `addons/l10n_ch/data/l10n_ch_chart_data.xml <{GITHUB_PATH}/addons/l10n_ch/data/l10n_ch_chart_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/l10n_ch_chart_data.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-at: l10nch_chart_template
    :end-at: </record>

.. note::

  Recommended **xmlid** for the record is `chart_template`.
  If you need many chart of accounts, you can add some suffixes, i.e. `chart_template_XXX`.

Chart of Accounts
=================

Account tags
------------

.. seealso::
   :ref:`Account Tag References <reference/account_account_tag>`

Tags are a way to sort accounts.
For example, imagine you want to create a financial report having multiple lines but you have no way to find a rule to dispatch the accounts according their ``code``.

> The solution is the usage of tags, one for each report line, to filter accounts like you want.

Reviewer 9: I know how rST works. This just looks schisophrenic.

Either the breaks are necessary for readability and they should also be in the rendered output, or they're not and they don't belong in the source. If the goal was to avoid overlong lines (which seems unlikely as the second line is >150) then it should be fill-paragraph'd.

It works with a prefix start/end, so every account where the code starts with something between start and end will have this account.group as parent group.  Furthermore, the account groups can have a parent account group as well to form the hierarchy.  

.. example::
  `addons/l10n_il/data/account.group.template.csv <{GITHUB_PATH}/addons/l10n_il/data/account.group.template.csv>`__.

  .. csv-table::
     :condition: odoo_dir_in_path
     :file: {ODOO_ABSPATH}/addons/l10n_il/data/account.group.template.csv
     :widths: 20,20,20,20,20
     :header-rows: 1

Taxes
-----

.. seealso::
   - :ref:`Tax References <reference/account_tax>`
   - :doc:`/applications/finance/accounting/taxation/taxes/taxes`

To add taxes you first need to specify tax groups. You normally need just one tax group for every tax rate, except for the 0% as you need to often distinguish between exempt, 0%, not subject, ... taxes. 
This model only has two required fields: name and country. Create file `data/account_tax_group_data.xml` and list the groups:

.. code-block:: xml

>     <?xml version="1.0" encoding="utf-8"?>

Reviewer 9: Leave out, it's unnecessary syntactic overhead (applies for every literal XML document)

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ch/data/account_tax_group_data.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-after: <data
    :end-before: </data>

.. example::
  `addons/l10n_uk/data/account.tax.group.csv <{GITHUB_PATH}/addons/l10n_uk/data/account.tax.group.csv>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_uk/data/account.tax.group.csv
    :condition: odoo_dir_in_path
    :language: csv

Now you can add the taxes via `data/account_tax_template_data.xml` file.  The first tax you define that is purchase/sale also becomes the default purchase/sale tax for your products.  

.. example::
  `addons/l10n_ae/data/account_tax_template_data.xml <{GITHUB_PATH}/addons/l10n_ae/data/account_tax_template_data.xml>`__.

  .. literalinclude:: {ODOO_RELPATH}/addons/l10n_ae/data/account_tax_template_data.xml
    :condition: odoo_dir_in_path
    :language: xml
    :start-at: uae_sale_tax_5_dubai
    :end-at: </record>

If some accounts should use default taxes, you can set them up in `data/account_account_template_post_data.xml`

Tax Report
----------

.. raw:: html

>    <div><span class="badge" style="background-color:#AD5E99">Enterprise feature</span><div>

Reviewer 9: Might be a good use-case for an `rst_epilog` and a "standard" (project-wide) substitution.

The tax report is declared in the Invoicing (`account`) app, but the report is only accessible when Accounting (`account_accountant`) is installed.

.. seealso::
   - :ref:`Tax Report Line References <reference/account_tax_report_line>`
   - :doc:`/applications/finance/accounting/reporting/declarations/tax_returns`

In the previous section you noticed fields `invoice_repartition_line_ids` / `refund_repartition_line_ids` and probably understood nothing about them. Good news: you are not alone on this incomprehension. Bad news: you have to figure it out a bit. The topic is complicated indeed:

.. graphviz:: images/tax_report.dot
    :class: overflow-auto

> The simple version is that in the tax template, you indicate in the invoice/refund repartition lines whether base or a percentage of the tax needs to be reported in which report line (through the minus/plus_report_line_ids).  It becomes clear also when you check the tax configuration in the Odoo interface (or check the docs (ref))

Reviewer 9: I don't understand a word of this. Also what's with the doubly-nested `(ref)`?
